native/src/main/java/io/ballerina/lib/data/csvdata/csv/CsvTraversal.java

```diff
@@ -0,0 +1,849 @@
+                            .map(mutableType -> CsvCreator.constructReadOnlyValue(
+                                    traverseCsv(csv, config, TypeCreator.createArrayType(mutableType))));
+                }
+            }
+            return Optional.empty();
+        }
+
+        private void traverseCsvWithUnionExpectedType(Type referredType, Type type, int sourceArraySize,
+                                                      BArray csv) {
+            for (Type memberType: ((UnionType) referredType).getMemberTypes()) {
+                Type mType = TypeUtils.getReferredType(memberType);
+                if (mType.getTag() == TypeTags.ARRAY_TAG) {
+                    int expectedArraySize = ((ArrayType) mType).getSize();
+                    resetForUnionTypes();
+                    try {
+                        setRootCsvNodeForNonUnionArrays(mType, mType);
+                        CsvUtils.validateExpectedArraySize(expectedArraySize, sourceArraySize);
+                        traverseCsvWithExpectedType(sourceArraySize, csv, type);
+                        return;
+                    } catch (Exception ex) {
+                        // ignore
+                    }
+                }
+            }
+            throw DiagnosticLog.error(DiagnosticErrorCode.SOURCE_CANNOT_CONVERT_INTO_EXP_TYPE, type);
+        }
+
+        private void traverseCsvWithExpectedType(int sourceArraySize,
+                                                 BArray csv, Type type) {
+            boolean isIntersection = false;
+            this.isFirstRowIsHeader = false;
+            if (expectedArrayElementType.getTag() == TypeTags.INTERSECTION_TAG) {
+                Optional<Type> mutableType = CsvUtils.getMutableType((IntersectionType) expectedArrayElementType);
+                if (mutableType.isPresent()) {
+                    isIntersection = true;
+                    expectedArrayElementType = mutableType.get();
+                }
+            }
+
+            switch (expectedArrayElementType.getTag()) {
+                case TypeTags.RECORD_TYPE_TAG:
+                case TypeTags.MAP_TAG:
+                    traverseCsvWithMappingAsExpectedType(sourceArraySize, csv,
+                            expectedArrayElementType, isIntersection);
+                    break;
+                case TypeTags.ARRAY_TAG:
+                case TypeTags.TUPLE_TAG:
+                    traverseCsvWithListAsExpectedType(sourceArraySize, csv,
+                            expectedArrayElementType, isIntersection);
+                    break;
+                case TypeTags.UNION_TAG:
+                    traverseCsvWithUnionExpectedType(csv,
+                            (UnionType) expectedArrayElementType, type);
+                    break;
+                default:
+                    throw DiagnosticLog.error(DiagnosticErrorCode.SOURCE_CANNOT_CONVERT_INTO_EXP_TYPE, type);
+            }
```

Comment:
I think we can use enhanced switch here

```suggestion
            switch (expectedArrayElementType.getTag()) {
                case TypeTags.RECORD_TYPE_TAG, TypeTags.MAP_TAG -> traverseCsvWithMappingAsExpectedType(sourceArraySize, csv,
                        expectedArrayElementType, isIntersection);
                case TypeTags.ARRAY_TAG, TypeTags.TUPLE_TAG -> traverseCsvWithListAsExpectedType(sourceArraySize, csv,
                        expectedArrayElementType, isIntersection);
                case TypeTags.UNION_TAG -> traverseCsvWithUnionExpectedType(csv,
                        (UnionType) expectedArrayElementType, type);
                default -> throw DiagnosticLog.error(DiagnosticErrorCode.SOURCE_CANNOT_CONVERT_INTO_EXP_TYPE, type);
            }
```